src/vulkan/command_buffer.cc

```diff
@@ -106,9 +96,28 @@ Result CommandBuffer::SubmitAndReset(uint32_t timeout_ms) {
   if (device_->GetPtrs()->vkResetCommandBuffer(command_, 0) != VK_SUCCESS)
     return Result("Vulkan::Calling vkResetCommandBuffer Fail");
 
-  state_ = CommandBufferState::kInitial;
   return {};
 }
 
+CommandBufferGuard::CommandBufferGuard(CommandBuffer* buffer) : buffer_(buffer) {
+  assert(!buffer_->guarded_);
+
+  buffer_->guarded_ = true;
+  result_ = buffer_->BeginRecording();
+}
+
+CommandBufferGuard::~CommandBufferGuard() {
+  assert(submitted_);
```

Comment:
This isn't the right thing to do if the flow has stopped before submission.
E.g. if any of the intermediate things fail with an error result, then this will assert out rather than report the error.

Reply:
Doh, fixed.